src/ngraph/builder/quantization/quantized_linear_matmul.cpp

```diff
@@ -0,0 +1,103 @@
+//*****************************************************************************
+// Copyright 2017-2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include "ngraph/builder/quantization/quantized_linear_matmul.hpp"
+#include "ngraph/axis_set.hpp"
+#include "ngraph/builder/make_constant.hpp"
+#include "ngraph/builder/quantization.hpp"
+#include "ngraph/op/constant.hpp"
+#include "ngraph/op/dequantize.hpp"
+#include "ngraph/op/divide.hpp"
+#include "ngraph/op/dot.hpp"
+#include "ngraph/op/experimental/quantized_dot.hpp"
+#include "ngraph/op/multiply.hpp"
+#include "ngraph/op/quantize.hpp"
+#include "ngraph/op/reshape.hpp"
+#include "ngraph/type/element_type.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+namespace ngraph
+{
+    namespace builder
+    {
+        namespace quantization
+        {
+            shared_ptr<Node> QuantizedLinearMatmul(shared_ptr<Node> input,
+                                                   shared_ptr<Node> filter,
+                                                   shared_ptr<Node> input_scale,
+                                                   shared_ptr<Node> filter_scale,
+                                                   shared_ptr<Node> output_scale)
+            {
+                auto requantization_scale = (input_scale * filter_scale) / output_scale;
+
+                // Since u8u8 goes to ref reshape the filter so that the ref dot can do a matmul
+                // which requires [m, n] * [n, k] order where as mkldnn requires [m, n] * [k, n]
+                if (input->get_element_type() == element::u8 &&
+                    filter->get_element_type() == element::u8)
+                {
+                    filter = make_shared<op::Reshape>(
```

Comment:
Can this reshape be done in the CPU backend?  We should not be encoding MKLDNN requirements at the op level.

Reply:
this check is more for ref kernel. I am reshaping it to [m, n] * [n, k] which is what ref expects. The onnx bridge will pass [m, n]* [k, n] which is what mkldnn expects. Assuming that the builders are CPU specific for now, I dont see a issue with this

Reply:
builders should not be cpu specific